src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,139 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    public static final int DEFAULT_CAPACITY = 10;
+    public static final double DEFAULT_CAPACITY_MULTIPLIER = 1.5;
+
+    private T[] container;
+    private int size;
+
+    public ArrayList() {
+        container = (T[])new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        growIfNecessary(1);
+        container[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index value while adding");
+        }
+        growIfNecessary(1);
+        insertElement(value, index);
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        growIfNecessary(list.size());
+        for (int i = 0; i < list.size(); i++) {
+            container[i + size] = list.get(i);
+        }
+        size += list.size();
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return container[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        validateIndex(index);
+        container[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        T removedElement = container[index];
+        removeElement(index);
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if (elementsAreEqual(element, container[i])) {
+                index = i;
+                break;
+            }
+        }
+        if (index == -1) {
+            throw new NoSuchElementException("Can't find element while removing");
+        }
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow() {
+        T []copy = (T[])new Object[(int)(container.length * DEFAULT_CAPACITY_MULTIPLIER)];
+        for (int i = 0; i < container.length; i++) {
+            copy[i] = container[i];
+        }
+        container = copy;
```

Comment:
rewrite this logic in next way:
1. calculate new capacity based on current array length and RESIZE_FACTOR
2. if array is full - resize it